Add option to apply filter actions to existing articles

raspo opened the pull request:
This PR adds a new checkbox to the `Filter actions` settings that lets the user apply filters retroactively; in other words: mark as read existing articles, not just new incoming ones.

<img width="1488" height="1328" alt="CleanShot 2025-09-13 at 18 15 34@2x" src="https://github.com/user-attachments/assets/47ce6e3c-5f92-49ee-abbc-6b6c96639a0a" />


**Background:**
I was trying to use the `Filter actions` to filter away certain type of articles that I never want to see in my feed (for example: sponsors or articles reposting the new podcast episode, etc).. but I was very confused by the fact that no matter what filter I'd try, existing articles would never get marked as read automatically. Eventually I discovered that this is intended behavior (as mentioned [here](https://github.com/FreshRSS/FreshRSS/issues/3152#issuecomment-769884867) and [here](https://github.com/FreshRSS/FreshRSS/issues/2941#issuecomment-623097799)).


Changes proposed in this pull request:

- Add new `Also apply filters to existing articles` checkbox
- Call `applyFiltersRetroactively()` when the checkbox has been checked
- `applyFiltersRetroactively` automatically marks as read each entry that is matching the filter

How to test the feature manually:

1. Select a feed from the sidebar, click the ⚙️ icon, then `Manage`
2. In the `Filter actions` section, add a filtering rule inside the `Mark an article as read…` field
3. Hit submit
4. Settings should be saved correctly
5. Open the feed settings again, but this time check the `Also apply filters to existing articles` checkbox
6. Hit submit again
7. Settings should be saved correctly and the filters should have been applied to existing articles (marking them read)

Pull request checklist:

- [x] clear commit messages
- [x] code manually tested
- [ ] unit tests written (optional if too hard)
- [ ] documentation updated

PS:
Thank you so much for your work on FreshRSS, I have been using it for many months and it has never needed any maintenance. I haven't written PHP in over 10 years, so I probably made a bunch of mistakes. For the translations I only added one string into the english one and then ran `make i18n-format`, I hope that's ok.

raspo committed:
Add option to apply filter actions to existing articles

Alkarex commented:
Thanks for the PR. Please run `make fix-all`

Alkarex committed:
Merge branch 'edge' into pr/raspo/7959

Alkarex committed:
make fix-all

Alkarex commented:
Side note: You are welcome to just push, and avoid force-push, which breaks many things: comments, reviews, permalinks, changes since last review, local branches, etc.

Alkarex committed:
Merge branch 'edge' into pr/raspo/7959

Alkarex committed:
Fixes

Alkarex commented:
Sorry I forgot a bit about this PR.
I am sightly worried regarding usability of this checkbox. What would you think of proposing a link that searches the corresponding articles to review them before marking as read?
I can make an example which can be reverted if that does not work well

raspo commented:
> Sorry I forgot a bit about this PR. I am sightly worried regarding usability of this checkbox. What would you think of proposing a link that searches the corresponding articles to review them before marking as read? I can make an example which can be reverted if that does not work well

No worries. That sounds good to me!

Alkarex committed:
Another approach with preview
Warning: does not work reliably with regex, since the DB's flavour is not necessarily the same than PHP's

Alkarex commented:
New approach: https://github.com/FreshRSS/FreshRSS/pull/7959/commits/20479475c90ac954b991b3703c3cc76c16aa2d5c

<img width="666" height="202" alt="image" src="https://github.com/user-attachments/assets/bb68ede4-60c8-4e0c-9317-c5ed7a6ad7df" />

Additional improvements:
* Also implemented at category level, and at global level
* Also implemented for favourites at global level

Shortcomings:
* Does not always work reliably with advanced regex, since the DB's flavour is not necessarily the same than PHP's

Tests and feedback welcome

Inverle commented:
I like it

Alkarex commented:
Related:
* https://github.com/FreshRSS/FreshRSS/pull/8141

Alkarex committed:
Merge branch 'edge' into pr/raspo/7959

Alkarex committed:
make fix-all

Alkarex commented:
@raspo Please add a line for you in https://github.com/FreshRSS/FreshRSS/blob/edge/CREDITS.md

Inverle commented:
#8222

Alkarex commented:
* https://github.com/FreshRSS/FreshRSS/pull/8259
